Improve CC documentation about CPU capacity and utilization (#5992)
strimzi/strimzi-kafka-operator: documentation/modules/cruise-control/proc-generating-optimization-proposals.adoc

----
+
The properties in the `Optimization Result` section describe the pending cluster rebalance operation. 
For descriptions of each property, see xref:contents-optimization-proposals[Contents of optimization proposals].

> .Insufficient capacity for cpu

Reviewer: ```suggestion
.Insufficient CPU capacity
```

Author: Yeah, that was actually taken from the error message. But I'll do the change.

> If the Kafka cluster is overloaded in terms of CPU utilization and needs some additional brokers, you may see an error like the following in the `KafkaRebalance` status. This may happen because, while proposals exclude the `excludedTopics`, the CPU utilization still considers replicas of the `excludedTopics`.

Author: It reads better, but it is not quite right.

> If the Kafka cluster is overloaded in terms of CPU utilization and needs some additional brokers, you may see an error like the following in the `KafkaRebalance` status. This may happen because, while proposals exclude the `excludedTopics`, the CPU utilization still considers replicas of the `excludedTopics`.

Author: The paragraph describes under which conditions you may see this error, so that you can verify if it matches with your setup. Instead, the note specifically addresses any doubts you may have looking at numbers. I would prefer the note to focus only on that aspect, as we know that it can be misleading.

> If the Kafka cluster is overloaded in terms of CPU utilization and needs some additional brokers, you may see an error like the following in the `KafkaRebalance` status. This may happen because, while proposals exclude the `excludedTopics`, the CPU utilization still considers replicas of the `excludedTopics`.

Reviewer 2: I felt that the second sentence starting with "This may happen because" makes it sound as if `excludedTopics` could be the cause of the brokers being overloaded. From what I understand, `excludedTopics` does not affect the load on the brokers, it affects what partitions are allowed to be moved?

I thought the mention of `excludeTopics` was to warn users who are expecting `excludeTopics` to affect the load that `excludeTopics` has no affect on the load. Was that the intention or am I misunderstanding something?

> If the Kafka cluster is overloaded in terms of CPU utilization and needs some additional brokers, you may see an error like the following in the `KafkaRebalance` status. This may happen because, while proposals exclude the `excludedTopics`, the CPU utilization still considers replicas of the `excludedTopics`.

Reviewer 2: I am fine with leaving the mention of `excludeTopics` up here in the first paragraph to prevent user confusion but I think we should rephrase so something like: 

> It's worth noting that this utilization value is unaffected by the excluded topics configuration (`excludedTopics`). Although optimization proposals will not reassign replicas of excluded topics, their load is still considered in the utilization calculation.

Also we will want one sentence per line in this file as well!

[source,shell,subs="+quotes"]
com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException:
        [CpuCapacityGoal] Insufficient capacity for cpu (Utilization 615.21,
        Allowed Capacity 420.00, Threshold: 0.70). Add at least 3 brokers with
        the same cpu capacity (100.00) as broker-0. Add at least 3 brokers with
        the same cpu capacity (100.00) as broker-0.

[NOTE]
====

> These numbers do not make sense if you look at `Kafka.spec.kafka.resources` because Strimzi generates a CPU percentage Cruise Control configuration, instead of a CPU core configuration. It is like having a single "virtual" CPU which has the cycles of 1 or more CPUs (the number configured in `Kafka.spec.kafka.resources.limits.cpu`). This is counterintuitive, but it has no effect on the rebalance behavior since the ratio between the utilization and capacity remains the same.

Reviewer: ```suggestion
The error suggests adding brokers with CPU capacity shown as a percentage rather than CPU cores. CPU `resources` are configured as integers representing cores in Strimzi. You can ignore the percentages. This has no effect on the rebalance behavior, since the ratio between CPU utilization and capacity remains the same.
```

Reviewer: I wasn't sure what were trying to say in this note. I found it difficult to relate to the error message. I've made a suggestion, assuming the main point is that we can ignore the numbers for capacity shown in the error.

Author: It reads better, but it is not quite right.

----
+
The properties in the `Optimization Result` section describe the pending cluster rebalance operation. 
For descriptions of each property, see xref:contents-optimization-proposals[Contents of optimization proposals].

.Insufficient CPU capacity

If a Kafka cluster is overloaded in terms of CPU utilization, you might see an insufficient CPU capacity error in the `KafkaRebalance` status. This might be because of excluded topics configuration (`excludedTopics`). Optimization proposals ignore excluded topics, but CPU utilization calculations consider their replicas.

.Example CPU utilization error
[source,shell,subs="+quotes"]
----
com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException:
        [CpuCapacityGoal] Insufficient capacity for cpu (Utilization 615.21,
        Allowed Capacity 420.00, Threshold: 0.70). Add at least 3 brokers with
        the same cpu capacity (100.00) as broker-0. Add at least 3 brokers with
        the same cpu capacity (100.00) as broker-0.
----

[NOTE]
====

> The error shows CPU capacity as a percentage rather than CPU cores. For that reason, it does not directly map to the number of CPUs configured in `Kafka.spec.kafka.resources`. It is like having a single "virtual" CPU per broker, which has the cycles of one or more CPUs. This has no effect on the rebalance behavior, since the ratio between CPU utilization and capacity remains the same.

Reviewer: ```suggestion
The error shows CPU capacity as a percentage rather than CPU cores. For that reason, it does not directly map to the number of CPUs configured in `Kafka.spec.kafka.resources`. It is like having a single _virtual_ CPU per broker, which has the cycles of one or more CPUs. This has no effect on the rebalance behavior, since the ratio between CPU utilization and capacity remains the same.
```

----
+
The properties in the `Optimization Result` section describe the pending cluster rebalance operation. 
For descriptions of each property, see xref:contents-optimization-proposals[Contents of optimization proposals].

.Insufficient CPU capacity

If a Kafka cluster is overloaded in terms of CPU utilization, you might see an insufficient CPU capacity error in the `KafkaRebalance` status. 

> This might be because of excluded topics configuration (`excludedTopics`). 

Reviewer 2: Adding comments from an "outdated" thread here for visibility:

I felt that this sentence makes it sound as if excludedTopics could be the cause of the brokers being overloaded. From what I understand, excludedTopics does not affect the load on the brokers, it affects what partitions are allowed to be moved?

I thought the mention of excludeTopics was to warn users who are expecting excludeTopics to affect the load that excludeTopics has no affect on the load. Was that the intention or am I misunderstanding something?

Author: I agree it is not clear enough.

AFAIK, when you set an `excludeTopics` regular expression, all topic names matching that expression are excluded from partition replica/leader movements in the optimization proposal. For that reason you, may never hit the CPU goal. Obviously, that would be a misconfiguration.

> Optimization proposals ignore excluded topics, but CPU utilization calculations consider their replicas.

Reviewer 2: I am fine with leaving the mention of excludeTopics up here in the first paragraph to prevent user confusion but I think we should rephrase so something like:

> It's worth noting that this utilization value is unaffected by the excluded topics configuration (excludedTopics). Although optimization proposals will not reassign replicas of excluded topics, their load is still considered in the utilization calculation.

Author: Much better, thanks!
